File: common/src/main/java/com/ssblur/alchimiae/events/reloadlisteners/EffectReloadListener.java
package com.ssblur.alchimiae.events.reloadlisteners;

import com.google.common.reflect.TypeToken;
import com.google.gson.JsonElement;
import com.ssblur.alchimiae.AlchimiaeMod;
import net.minecraft.resources.ResourceLocation;

import java.lang.reflect.Type;
import java.util.HashMap;

public class EffectReloadListener extends AlchimiaeReloadListener {
  public record EffectResource(String effect, float rarity){}
  static Type EFFECT_TYPE = new TypeToken<EffectResource>() {}.getType();
  public static final EffectReloadListener INSTANCE = new EffectReloadListener();

  public HashMap<ResourceLocation, EffectResource> effects = new HashMap<>();

  public EffectReloadListener() {
    super("alchimiae/effects");
  }

  @Override
  public void loadResource(ResourceLocation resourceLocation, JsonElement jsonElement) {
    EffectResource resource = GSON.fromJson(jsonElement, EFFECT_TYPE);
    AlchimiaeMod.LOGGER.debug(
      "Loaded effect {} with potion effect {} and rarity {}",
      resourceLocation,
      resource.effect,
      resource.rarity
    );
    effects.put(resourceLocation, resource);
  }
}

File: common/src/main/kotlin/com/ssblur/alchimiae/events/reloadlisteners/EffectReloadListener.kt
package com.ssblur.alchimiae.events.reloadlisteners

import com.google.common.reflect.TypeToken
import com.google.gson.JsonElement
import com.ssblur.alchimiae.AlchimiaeMod
import net.minecraft.resources.ResourceLocation
import java.lang.reflect.Type

class EffectReloadListener : AlchimiaeReloadListener("alchimiae/effects") {
  @JvmRecord
  data class EffectResource(val effect: String, val rarity: Float)

  var effects: HashMap<ResourceLocation, EffectResource> = HashMap()

  override fun loadResource(resourceLocation: ResourceLocation, jsonElement: JsonElement?) {
    val resource: EffectResource =
      AlchimiaeReloadListener.Companion.GSON.fromJson<EffectResource>(jsonElement, EFFECT_TYPE)
    AlchimiaeMod.LOGGER.debug(
      "Loaded effect {} with potion effect {} and rarity {}",
      resourceLocation,
      resource.effect,
      resource.rarity
    )
    effects[resourceLocation] = resource
  }

  companion object {
    var EFFECT_TYPE: Type = object : TypeToken<EffectResource?>() {}.type
    val INSTANCE: EffectReloadListener = EffectReloadListener()
  }
}
